File: client/src/types/property.ts
export type PropertyType = 'apartment' | 'villa' | 'penthouse' | 'plot' | 'commercial' | 'studio';

export type PropertyStatus = 'available' | 'reserved' | 'booked' | 'sold';

export interface PropertyImage {
  id: number;
  url: string;
  caption?: string;
  is_primary: boolean;
  sort_order: number;
}

export interface Builder {
  id: number;
  name: string;
  logo_url?: string;
  description?: string;
  rating: number;
  total_projects: number;
  delivered_projects: number;
  headquarters?: string;
  established_year?: number;
  is_verified: boolean;
}

export interface City {
  id: number;
  name: string;
  state: string;
}

export interface PropertyListItem {
  id: number;
  name: string;
  slug: string;
  property_type: PropertyType;
  status: PropertyStatus;
  locality?: string;
  city?: string;
  builder_name?: string;
  bedrooms?: number;
  bathrooms?: number;
  area_sqft?: number;
  price: number;
  price_per_sqft?: number;
  possession_date?: string;
  rating: number;
  review_count: number;
  is_featured: boolean;
  primary_image?: string;
  expected_roi?: number;
  is_wishlisted?: boolean;
}

export interface PropertyDetail {
  id: number;
  name: string;
  slug: string;
  description?: string;
  property_type: PropertyType;
  status: PropertyStatus;
  locality?: string;
  full_address?: string;
  latitude?: number;
  longitude?: number;
  pincode?: string;
  bedrooms?: number;
  bathrooms?: number;
  area_sqft?: number;
  total_floors?: number;
  floor_number?: number;
  parking_spots: number;
  facing?: string;
  furnishing?: string;
  price: number;
  price_per_sqft?: number;
  maintenance_monthly?: number;
  possession_date?: string;
  rating: number;
  review_count: number;
  view_count: number;
  is_featured: boolean;
  is_verified: boolean;
  rera_number?: string;
  expected_roi?: number;
  created_at: string;
  carpet_area?: number;
  builtup_area?: number;
  project_name?: string;
  highlights?: { icon: string; title: string; description: string }[];
  nearby_locations?: { category: string; name: string; distance: string }[];
  documents?: { title: string; type: string; size: string; category: string; download_url: string }[];
  images: PropertyImage[];
  amenity_names: string[];
  builder?: Builder;
  city?: City;
  is_wishlisted?: boolean;
}

export interface PropertyFilterParams {
  city?: string;
  min_price?: number;
  max_price?: number;
  bedrooms?: number;
  bathrooms?: number;
  property_type?: string;
  builder_id?: number;
  status?: string;
  min_area?: number;
  max_area?: number;
  search?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  page?: number;
  limit?: number;
}

export interface PropertyListResponse {
  items: PropertyListItem[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

File: server/schemas/property.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class PropertyImageOut(BaseModel):
    id: int
    url: str
    caption: Optional[str] = None
    is_primary: bool = False
    sort_order: int = 0

    model_config = {"from_attributes": True}


class AmenityOut(BaseModel):
    id: int
    name: str
    icon: Optional[str] = None
    category: Optional[str] = None

    model_config = {"from_attributes": True}


class BuilderOut(BaseModel):
    id: int
    name: str
    logo_url: Optional[str] = None
    rating: float
    total_projects: int
    delivered_projects: int
    headquarters: Optional[str] = None
    established_year: Optional[int] = None
    is_verified: bool = True

    model_config = {"from_attributes": True}


class CityOut(BaseModel):
    id: int
    name: str
    state: str

    model_config = {"from_attributes": True}


class PropertyListItem(BaseModel):
    id: int
    name: str
    slug: str
    property_type: str
    status: str
    locality: Optional[str] = None
    city: Optional[str] = None
    builder_name: Optional[str] = None
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    area_sqft: Optional[float] = None
    price: int
    price_per_sqft: Optional[float] = None
    possession_date: Optional[str] = None
    rating: float
    review_count: int
    is_featured: bool
    primary_image: Optional[str] = None
    expected_roi: Optional[float] = None

    model_config = {"from_attributes": True}


class PropertyDetail(BaseModel):
    id: int
    name: str
    slug: str
    description: Optional[str] = None
    property_type: str
    status: str
    locality: Optional[str] = None
    full_address: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    pincode: Optional[str] = None
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    area_sqft: Optional[float] = None
    total_floors: Optional[int] = None
    floor_number: Optional[int] = None
    parking_spots: int = 1
    facing: Optional[str] = None
    furnishing: Optional[str] = None
    price: int
    price_per_sqft: Optional[float] = None
    maintenance_monthly: Optional[int] = None
    possession_date: Optional[str] = None
    rating: float
    review_count: int
    view_count: int
    is_featured: bool
    is_verified: bool
    rera_number: Optional[str] = None
    expected_roi: Optional[float] = None
    created_at: datetime
    images: List[PropertyImageOut] = []
    amenity_names: List[str] = []
    builder: Optional[BuilderOut] = None
    city: Optional[CityOut] = None

    model_config = {"from_attributes": True}


class PropertyFilter(BaseModel):
    city: Optional[str] = None
    min_price: Optional[int] = None
    max_price: Optional[int] = None
    bedrooms: Optional[int] = None
    bathrooms: Optional[int] = None
    property_type: Optional[str] = None
    builder_id: Optional[int] = None
    status: Optional[str] = None
    min_area: Optional[float] = None
    max_area: Optional[float] = None
    search: Optional[str] = None
    sort_by: Optional[str] = "created_at"
    sort_order: Optional[str] = "desc"
    page: int = 1
    limit: int = 12


class PropertyListResponse(BaseModel):
    items: List[PropertyListItem]
    total: int
    page: int
    limit: int
    pages: int
